lib/dialects/abstract/query-generator/query-proto.js

```diff
@@ -0,0 +1,156 @@
+'use strict';
+
+const { Composition } = require('./composition');
+
+/*
+ * Named collections of Composition elements.
+ *
+ * Useful for organizing and manipulating the different parts of a query.
+ * Allow disordered query building and easy substitution of query parts. Flexible,
+ * non-nested and non-exahustive organization.
+ */
+
+class QueryProto {
+  constructor(proto) {
+    if (proto) {
+      this.set(proto);
+    } else {
+      // Should clone Composition, NOT copy shallow references
+      for (const key of this.constructor.partNames) {
+        this[key] = new Composition();
+      }
+    };
+  }
+  set(proto) {
+    // Should clone proto into this object, NOT copy shallow references
+
+    for (const key of this.constructor.partNames) {
+      this[key] = new Composition();
+
+      if (proto[key] && proto[key].length) {
+        this[key] = proto[key].clone();
+      } else {
+        this[key] = new Composition();
+      }
+    }
+
+    return this;
+  }
+  clone() {
+    // Should clone Composition, NOT copy shallow references
+    const clone = new this.constructor();
+
+    for (const key of this.constructor.partNames) {
+      clone[key] = this[key].clone();
+    }
+
+    return clone;
+  }
+  toComposition() {
+    throw new Error(`Class ${this.constructor.name} has not implemented toComposition method`);
+  }
+}
+
+QueryProto.partNames = [];
```

Comment:
You can define it as a static getter in class itself, for clarity

Reply:
Good idea